### DISCUSSION TITLE
State Update Issue in Composable Navigation

### DISCUSSION THREAD

**Discussion by mlouli** [pointfreeco/swift-composable-architecture#2309] [Q&A] [answered]
Hello, I've got a question about the implementation of composable navigation.

I implemented a complete flow as mentioned in this [composable navigation post](https://www.pointfree.co/blog/posts/104-composable-navigation-beta). I grouped all the possible sheets my app can present into a single view, which we'll call the 'master sheet'. This sheet has an enum of all possible child sheets as a state property.

All these child sheets utilize TCA.

**My issue is**, even after implementing the technique from the article, when a child sheet updates its state, the view doesn't register this change. But when the same child sheet is presented directly from the parent, everything operates smoothly.

Could this be a limitation, bug, or am I possibly making a mistake?

Thank you in advance.

**Comment by mbrandonw** [comment]
Hi @mlouli, first I would suggest following the [tutorials](https://pointfreeco.github.io/swift-composable-architecture/main/tutorials/meetcomposablearchitecture) in the repo rather than that blog post, which was posted many months ago at the beginning of the beta period.

And second, it's hard to say what the problem is without seeing some code. Can you share a minimal project that demonstrates the problem you are seeing?

**Reply by mlouli** [reply]
Hello @mbrandonw, appreciate your swift response 🏎️!

The tutorials are indeed comprehensive and addressed my concerns, particularly the [Navigation/Multiple presentation destinations](https://pointfreeco.github.io/swift-composable-architecture/main/tutorials/composablearchitecture/02-02-multipledestinations).

Regarding my second issue, I found that the problem isn't what I initially thought. I noticed that combining Vanilla SwiftUI and TCA Navigation introduces unusual behaviour. I'm curious if there's a way to prevent this.

To illustrate, consider the root presenting a child sheet, which is TCA-based and changes color based on the isOn state. This child sheet is wrapped in a navigation view (the modal wrapper), which has a 'done' button to dismiss the view.

In the modal wrapper, I've used `@Environment(\.dismiss) private var dismiss`, which is causing the problem. This results in the child view not registering state changes, hence the color doesn't change from red to green.

Can there be warnings to indicate this, or should it be considered a developer oversight?

<details open>
<summary>Demo</summary>
<br>

```swift
import SwiftUI
import ComposableArchitecture

@main
struct AppScene: SwiftUI.App {
	var body: some Scene {
		WindowGroup {
			RootView(store: .init(
				initialState: .init(),
				reducer: Root()
			))
		}
	}
}

// MARK: - Root

struct Root: ReducerProtocol {
	struct State: Equatable {
		@PresentationState var sheet: AnySheet.State?
	}

	enum Action: Equatable {
		case presentSheetButtonTapped
		case sheet(PresentationAction<AnySheet.Action>)
	}

	var body: some ReducerProtocolOf<Self> {
		Reduce { state, action in
			switch action {
			case .presentSheetButtonTapped:
				state.sheet = .child
				return .none
			case .sheet:
				return .none
			}
		}
		.ifLet(\.$sheet, action: /Action.sheet) {
			AnySheet()
		}
	}
}

struct RootView: View {
	let store: StoreOf<Root>

	var body: some View {
		WithViewStore(self.store.stateless) { viewStore in
			Button("Present Child") {
				viewStore.send(.presentSheetButtonTapped)
			}
		}
		.sheet(store: self.store.scope(
			state: \.$sheet,
			action: { .sheet($0) }
		)) { store in
			AnySheetView(store: store)
		}
	}
}

// MARK: - AnySheet

struct AnySheet: ReducerProtocol {
	enum State: Equatable {
		case child
	}

	enum Action: Equatable {}

	func reduce(into state: inout State, action: Action) -> EffectTask<Action> {}
}

struct AnySheetView: View {
	let store: StoreOf<AnySheet>

	var body: some View {
		WithViewStore(self.store, observe: { $0 }) { viewStore in
			switch viewStore.state {
			case .child:
				ModalWrapper {
					ChildView(store: .init(
						initialState: .init(),
						reducer: Child()
					))
				}
			}
		}
	}
}

// MARK: - Child

struct Child: ReducerProtocol {
	struct State: Equatable {
		var isOn = false
	}

	enum Action: Equatable {
		case onAppear
	}

	func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
		switch action {
		case .onAppear:
			state.isOn = true
			return .none
		}
	}
}

struct ChildView: View {
	let store: StoreOf<Child>

	var body: some View {
		WithViewStore(self.store, observe: \.isOn) { viewStore in
			ZStack {
				if viewStore.state {
					Color.green
				} else {
					Color.red
				}
			}
			.edgesIgnoringSafeArea(.all)
			.task {
				try! await Task.sleep(seconds: 1)
				viewStore.send(.onAppear)
			}
		}
	}
}

// MARK: - Misc

struct ModalWrapper<Content: View>: View {
	private let content: () -> Content
	@Environment(\.dismiss) private var dismiss

	init(@ViewBuilder content: @escaping () -> Content) {
		self.content = content
	}

	var body: some View {
		NavigationView {
			content()
				.toolbar {
					ToolbarItem {
						Button {
//							dismiss()
						} label: {
							Text("Done")
								.fontWeight(.semibold)
						}
					}
				}
		}
	}
}
```
</details>

**Reply by mbrandonw** [reply]
This demo works just fine for me when running in Xcode 14.3.1 and iOS 16.4. The sheet opens, the background goes from red to green, and then tapping "Done" dismisses.

**Reply by mlouli** [reply]
I have a similar setup with swift-composable-architecture 0.54.1. Upgrading to v0.56.0 didn't resolve the issue. I'm also using swiftui-navigation 0.8.0.

https://github.com/pointfreeco/swift-composable-architecture/assets/14870254/9fb38e70-bcdb-482b-a957-466e50dcc180

**Reply by mbrandonw** [reply]
Can you upload a project that demonstrates the problem? I just copied and pasted your code into an existing project.

**Reply by mlouli** [reply]
Sorry for the delay; I needed to sanitize the project data. Surprisingly, the identical code runs perfectly in a fresh project.

The shared project should illustrate the problem.

[Debug.zip](https://github.com/pointfreeco/swift-composable-architecture/files/12099565/Debug.zip)

**Reply by mbrandonw** [reply, marked as answer]
I didn't notice in your first code snippet, but the problem is that you are creating a whole new store from scratch to pass along to the child view:

```swift
      case .child:
        ModalWrapper {
          ChildView(store: .init(
            initialState: .init(),
            reducer: Child()
          ))
        }
```

That store is fully disconnected from the parent store, and so it is not surprising that it isn't picking up state changes. Typically one derives a store from the parent to hand down to the child.

It seems that either that state should not be an enum, or you should follow the steps outlined in the tutorial (in particular [this section](https://pointfreeco.github.io/swift-composable-architecture/main/tutorials/composablearchitecture/02-02-multipledestinations)) to see how to properly drive navigation from an enum of destinations.

**Reply by mlouli** [reply]
Thanks for your quick responses; the tutorial section you pointed out is indeed quite clear. 

Enjoy the rest of your day!
